modules/dnn/src/layers/layer_norm.cpp

```diff
@@ -0,0 +1,160 @@
+class LayerNormLayerImpl CV_FINAL : public LayerNormLayer
+{
+public:
+    LayerNormLayerImpl(const LayerParams& params)
+    {
+        setParamsFrom(params);
+
+        // standard attr
+        axis = params.get<int>("axis", 0);
+        epsilon = params.get<float>("epsilon", 1e-5);
+
+        // opencv attr
+        hasBias = params.get<bool>("hasBias", false);
+    }
+
+    virtual bool supportBackend(int backendId) CV_OVERRIDE
+    {
+        return backendId == DNN_BACKEND_OPENCV;
+    }
+
+    virtual bool getMemoryShapes(const std::vector<MatShape> &inputs,
+                                 const int requiredOutputs,
+                                 std::vector<MatShape> &outputs,
+                                 std::vector<MatShape> &internals) const CV_OVERRIDE
+    {
+        // check shapes of weight and bias if existed
+        // inputs >= 2 (X and Weight are requested, bias is optional)
+        CV_CheckGE(inputs.size(), 2ull, "LayerNorm: require at least two inputs (x, weight)");
+        CV_CheckLE(inputs.size(), 3ull, "LayerNorm: have at most three inputs (x, weight, bias)");
+
+        auto x_shape = inputs[0];
+        int x_ndims = static_cast<int>(x_shape.size());
+
+        auto w_shape = inputs[1];
+        // if axis == last_dim, scale and b are both 1d tensor (represented as 2d mat nx1)
+        int w_ndims = static_cast<int>(w_shape.size());
+        w_ndims = (axis == x_ndims - 1 && w_ndims == 2) ? w_ndims - 1 : w_ndims;
+        CV_CheckEQ(x_ndims - axis, w_ndims, "LayerNorm: shape of weight does not match with given axis and shape of input");
+        for (int i = 0; i < w_ndims; ++i)
+            CV_CheckEQ(x_shape[axis+i], w_shape[i], "LayerNorm: weight dimensions does not match with input dimensions");
+        if (hasBias)
+        {
+            auto b_shape = inputs[2];
+            CV_CheckEQ(w_shape.size(), b_shape.size(), "LayerNorm: shape of weight does not match with shape of bias");
+            for (size_t i = 0; i < w_shape.size(); ++i)
+                CV_CheckEQ(w_shape[i], b_shape[i], "LayerNorm: bias dimensions does not match with weight dimensions");
+        }
+
+        // only one output is needed; Mean & InvStdDev are not needed
+        // in inference and should beomitted in onnx importer
+        outputs.assign(1, inputs[0]);
+        return false;
+    }
+
+    class LayerNormInvoker : public ParallelLoopBody
+    {
+    public:
+        const Mat* src, *scale, *bias;
+        Mat *dst;
+
+        float epsilon;
+        int nstripes;
+
+        int total;
+        int stripeSize;
+        int normSize;
+
+        LayerNormInvoker() : src(0), scale(0), bias(0), dst(0), epsilon(0.f), nstripes(0), total(0), stripeSize(0), normSize(0) { }
+
+        static void run(const Mat* src, const Mat* scale, const Mat* b, Mat* dst, int axis, float epsilon, int nstripes)
+        {
+            CV_Assert_N( src->isContinuous(), dst->isContinuous(), src->type() == CV_32F, src->type() == dst->type());
```

Comment:
OpenCL tests failed:

```
layer_norm.cpp:81: error: (-215:Assertion failed) src->type() == 5 in function 'run'
" thrown in the test body.
[  FAILED  ] Test_ONNX_layers.LayerNorm/1, where GetParam() = OCV/OCL_FP16 (1 ms)
```

Reply:
@alalek Tests on `OCL_FP16` are failed because source type is limited to `CV_32F`. The reason of the limitation is I use float as data type for calculation implementation. Any suggestions?

Reply:
BTW,

> `== 5`

Avoid using of `CV_Assert_N` as its implementation is buggy and eats original names (CV_32F): #12219
Use `CV_Assert` or `CV_CheckType*` (preferable) macros.

Reply:
> The reason of the limitation is I use float as data type for calculation implementation. Any suggestions?

Similar to other layers - there is a fallback code with conversions: `FP16 -> FP32 -> compute as FP32 -> FP16`.

```
        if (inputs_arr.depth() == CV_16S)
        {
            forward_fallback(inputs_arr, outputs_arr, internals_arr);
            return;
        }
```

(CV_16S represents CV_FP16)

Reply:
Thanks for all the suggestions and explanation! Updated a commit to solve these issues hopefully.